src/crimson/osd/watch.cc

```diff
@@ -66,6 +66,52 @@ seastar::future<> Watch::notify_ack(
   });
 }
 
+seastar::future<> Watch::send_disconnect_msg()
+{
+  if (!is_connected()) {
+    return seastar::now();
+  }
+  ceph::bufferlist empty;
+  return conn->send(make_message<MWatchNotify>(
+    winfo.cookie,
+    0,
+    0,
+    CEPH_WATCH_EVENT_DISCONNECT,
+    empty));
+}
+
+void Watch::discard_state()
+{
+  ceph_assert(obc);
+  in_progress_notifies.clear();
+}
+
+seastar::future<> Watch::remove(const bool send_disconnect)
+{
+  logger().info("{}", __func__);
+  auto disconnected = send_disconnect ? send_disconnect_msg()
+                                      : seastar::now();
```

Comment:
off topic, the pattern is repeating itself, i am wondering if we should have something like
```c++
template<typename F>
auto do_if(F&& func, bool predicate) {
  if (predicate) {
    return seastar::make_ready_future();
  } else {
    return std::forward<Func>(func)();
  }
}
```
either in crimson or in seastar? what do you think?

Reply:
I would vote for staging in our `crimson/common` first. Then, if turns useful, let's upstream it.